### PR TITLE
[155] encodedsession: Support ResultSet with a huge header

### PR DESCRIPTION
**Pull request by madscientist**
The ResultSet consists of a header (labels for each returned column)
plus some number of rows.  The maximum size of a single "fetch"
operation is 100,000 bytes.  If the header happens to exceed that size
then the fetch operation will always raise an EndOfStream exception.

The protocol for result sets is a little fragile: we send back a row of
data followed by an optional integer.  If the integer is 1 then there is
another row to be read in this same fetch result message.  If the
integer is 0 then there are no more rows to fetch (the ResultSet is
complete).  If the integer is missing completely then there are more
rows to fetch by sending another request to the server.

In the driver we were always assuming there would be the headers plus at
least one row before we ran out of room in the message, but if the
header was large enough (> 100,000 bytes for labels) that might not be
true.

Add a new encodedsession._hasBytes() method that allows us to check
whether there is data left in the message.  Use this to rewrite the
fetch_result_set() and fetch_result_set_next() methods to simplify the
logic and avoid the error above.

### CODE REVIEW

File: pynuodb/encodedsession.py [hunk 1]
```diff
@@ -438,30 +438,27 @@ def fetch_result_set(self, statement):
         handle = self.getInt()
         colcount = self.getInt()
 
-        col_num_iter = range(colcount)
-        for _ in col_num_iter:
+        # skip the header labels
+        for _ in range(colcount):
```
Comment (madscientist) [inline pynuodb/encodedsession.py:442]: I'm not sure why this is done, but anyway I didn't change it.

Comment (mkysel) [reply pynuodb/encodedsession.py:442]: where is the change coming from then? Does it work?

**Review by mkysel — APPROVED** [review]
looks good

File: pynuodb/encodedsession.py (hunk 1 above)
Comment (mkysel) [reply pynuodb/encodedsession.py:442]: got it, it sounded like you did not know where the change came from :)

**Pull request merged**